### PR TITLE
Jim/lineage design

### PR DESCRIPTION
**Pull request by jimduff-usds**
This PR is the design proposal for the base tables tracking lineage and history of Reports and Items in the web.

Its easier to read the proposal here: https://github.com/CDCgov/prime-data-hub/tree/jim/lineage-design/prime-router/docs/proposals/lineage

## Designs

- #147 

## To Be Done

- The actual implementation of #147

### CODE REVIEW

File: prime-router/docs/proposals/lineage/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+## Assumptions
+
+- TASKs create zero or more REPORTs which have one or more ITEMs
+- ITEMs are like quarks.They only exist inside of a REPORT.  Its fine and quite common to have a report with one item.
+- We must track lineage per-item, however, people think in terms of REPORTs.  Hence we need to track both.
+- Tracking of sending- and receiving- Organizations is crucial, to know who gets permissions to see the data.
+- Many:many -- A given ACTION could take in many reports, and produce many reports.  Need to allow for both splitting and merging.
+- Many:many -- An ITEM can be the result of multiple parent items, and can result in multiple child ITEMs.
+- A few schema best practices:  Every table must have a single col PK (either a serial or a UUID), and a created_date, Normalize independent data.   Use singular nouns.
```
Comment (ronaldheft-usds) [inline prime-router/docs/proposals/lineage/README.md:58]: Agree with these assumptions.

File: prime-router/docs/proposals/lineage/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+- Tracking versions of the functions that created the data, and SHA hash of every report.
+- Next steps: Have an easy-to-use API that delivers history
+- Next steps: Have an easy-to-use API that can retrieve the data in an individual ITEM. (?)
+
+### Rollout plan:
+
+1.  Build out the new schema in parallel with the existing TASK, TASK_SOURCE tables.
+2.  Wire the REPORT_FILE and REPORT_LINEAGE tables first, then the ITEM and ITEM_LINEAGE.
+3.  Code into the validation step a requirement for uniqueness of tracking Ids within each Received report. 
+4.  Migrate existing data to the new tables, and retire or modify the old ones.
```
Comment (ronaldheft-usds) [inline prime-router/docs/proposals/lineage/README.md:92]: A few questions on the rollout plan:

* What do you mean by built in parallel? Will the schema be deployed with no data until we're ready to migrate? Will data exist in both schemas for a period of time?
* When we do a cut over to the new schema, how do we plan out migrating the old data? Will there be a migration process that copies the data over time? A one-off script that runs (with I assume some system downtime)?
* What's the amount of data we're looking to migrate?

**Review by ronaldheft-usds — COMMENTED** [review]
Still getting up to speed on everything, so I don't feel comfortable commenting on the schema design, but I do have some questions on the rollout plan.

File: prime-router/docs/proposals/lineage/README.md (hunk 2 above)
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:92]: @ronaldheft-usds Good questions - thanks!   I've modified the README, but here's the quick answers:
- data will exist in both for a period of time.
- we'll do a quick one-off script.
- Tiny amount of data.   less than 1000 rows at the moment.

File: prime-router/docs/proposals/lineage/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+- Many:many -- A given ACTION could take in many reports, and produce many reports.  Need to allow for both splitting and merging.
+- Many:many -- An ITEM can be the result of multiple parent items, and can result in multiple child ITEMs.
+- A few schema best practices:  Every table must have a single col PK (either a serial or a UUID), and a created_date, Normalize independent data.   Use singular nouns.
+-- Receipt of a dataset implies the right to see the lineage of that dataset.
+-- Sending a dataset implies the right to see the lineage of that dataset. 
+
+## Proposed Design
+
+Please [pull up the ER diagram](./lineage-er-diagram.jpeg) for the detailed proposed schema.
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:65]: I've been using this SQL style guide https://about.gitlab.com/handbook/business-ops/data-team/platform/sql-style-guide. Read it over, if you agree with its recommendations, then we can check it in. One thing I noticed in your schema was the use of `_date` vs ` _at` for timestamps.

File: prime-router/docs/proposals/lineage/README.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+| Receive   | Receive a new Report into the Hub
+| Transform | Validate a received report, and transform into desired output formats, according to configured data filter business rules
+| Batch     |  Merge one or more reports together.   Typically done to consolidate data per receiver requirements.
+| Send      | Reliably delivery data to an external receiver, eg, an API, sftp site, or provide for upload (and notify with an email)
+| Wipe      | Remove data, per business rules
+
+An **Organization** can recieve and/or send data to the hub.  A typical Organization is a state or local Public Health Department.   Strac, Redox, and SimpleReport are all also Organizations in the hub.
+
+An **OrganizationService** is an endpoint to which the Hub can send a specific kind of data to a specific place, so that the Organiation can use the data.  For example, an OrganizationService might be an sftp site where a local health department wants hl7 files uploaded every 4 hours.
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:46]: In my proposal to come, I'm thinking about going back to Senders and Receivers to OrganizationClient and OrganizationService. Perhaps, if you agree, then you can make the same change.

File: prime-router/docs/proposals/lineage/README.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+### How this design helps build trust
+
+- Fine-grained lineage tracking at the ITEM level, with ability to query both "backward" and "forward"
+- Improved tracking of sender and receiver ownership
+- Tracking versions of the functions that created the data, and SHA hash of every report.
+- Next steps: Have an easy-to-use API that delivers history
+- Next steps: Have an easy-to-use API that can retrieve the data in an individual ITEM. (?)
+
+### Rollout plan:
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:88]: When you say rollout plan, I assuming that you talking about separate PRs that could go to production? I wonder what the operational implications of having two sets of tables going in parallel means? I suggest that the two at once state will be somewhat problematic. If so, could we simplify this plan into fewer steps that executed faster? We can take advantage of the fact that we have very low transaction rate now.

File: prime-router/docs/proposals/lineage/README.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+2.  Wire the REPORT_FILE and REPORT_LINEAGE tables first, then the ITEM and ITEM_LINEAGE.
+3.  Code into the validation step a requirement for uniqueness of tracking Ids within each Received report. 
+4.  Migrate existing data to the new tables, and retire or modify the old ones.   Migrate all at once, as a one-off, with a brief downtime.  (Note:  I believe some of the data in the TASK_SOURCE table is not correct, so may only wwant to migrate correct data.)   Right now there are only a few hundred rows of data to move.
+
+
+#### Queries
+
+To validate that we can design recursive parent/child queries, I experimented with a much simplified schema found here:
+[recursive-parent-child-test.sql](recursive-parent-child-test.sql)
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:99]: Thank you for including these queries.

File: prime-router/docs/proposals/lineage/README.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+### Notes on the Proposed Schema
+
+- The separation of Task from Report allows us to have Tasks that don't necessarily create Reports.
+- Addition of the `TASK.task_result` fills a gap where we're not tracking the error messages sent back to users.
+- Apparently I can't name the REPORT table 'REPORT', due to a limitation in Flyway/jooq.  Hence the weird name `REPORT_FILE`.   I also thought about `REPORT_IN`, `REPORT_DATA`, `REPORT_REC`.  Haven't sunk as low as `REPURT`, but I'm on my way there...
+- Many values in REPORT_FILE will be empty depending on the task creating the report.  For example, all the sending_* fields are only filled in for an initially received report.   All the receiving_* fields are only filled in for reports sent externally.
+- might need a report_status in the future.  For example, `wiped` or `archived`, to explain why a blob might have disappeared.
+- The primary key to use for ITEM is tricky.  The sender gives us the `tracking_element`, but we have little control over how robust that is.  Because of this, all queries on `tracking_element` must also filter on report_id.
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:75]: Agreed. Maybe we should manufacture a tracking_element ID if one is not provided by the sender?

File: prime-router/docs/proposals/lineage/README.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+- A few schema best practices:  Every table must have a single col PK (either a serial or a UUID), and a created_date, Normalize independent data.   Use singular nouns.
+-- Receipt of a dataset implies the right to see the lineage of that dataset.
+-- Sending a dataset implies the right to see the lineage of that dataset. 
+
+## Proposed Design
+
+Please [pull up the ER diagram](./lineage-er-diagram.jpeg) for the detailed proposed schema.
+
+(Note: [Original lucid chart diagram is here](https://lucid.app/lucidchart/7b81cccb-91a0-44b3-b17f-90b9b0c8304e/edit?beaconFlowId=332110D5160D6847&page=0_0#))
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:67]: I would like to track the `receiving_org_results` and `receiving_org_params` like you are proposing for the sending side. Redox, in particular, sends back an ID that I want to track in our DB.

File: prime-router/docs/proposals/lineage/README.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+- Build trust in the Hub - the schema should enable features that give both receivers and senders a sense of trust in the system.  The hub should be able to reassure our users that their data has transferred safely and securely.
+- Secure data at all steps, both in-transit and at rest, while still making it available to those who are authorized access.
+
+### Future Use Cases
+
+For the first rounds of work, I'm avoiding tracking any specific data *internal* to a report, other than the `tracking_elements`.   However, the schema is designed to be expanded in future sprints, to potentially accomodate these use cases:
+
+- Tracking Positivity rates, sharp increases/drops for specific senders, geographic spikes, etc.
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/lineage/README.md:26]: Cool.

**Review by RickHawesUSDS — APPROVED** [review]
Great job. 

Please read the comments. There are a few side issues on the SQL style and the rollout plan which I would appreciate a conversation on.

File: prime-router/docs/proposals/lineage/README.md (hunk 3 above)
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:65]: Switched to `_at`.

File: prime-router/docs/proposals/lineage/README.md (hunk 8 above)
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:67]: I've added this.

File: prime-router/docs/proposals/lineage/README.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+## Proposed Design
+
+Please [pull up the ER diagram](./lineage-er-diagram.jpeg) for the detailed proposed schema.
+
+(Note: [Original lucid chart diagram is here](https://lucid.app/lucidchart/7b81cccb-91a0-44b3-b17f-90b9b0c8304e/edit?beaconFlowId=332110D5160D6847&page=0_0#))
+
+### Notes on the Proposed Schema
+
+- The separation of Task from Report allows us to have Tasks that don't necessarily create Reports.
+- Addition of the `TASK.task_result` fills a gap where we're not tracking the error messages sent back to users.
```
[the remark this answers is not in the document]
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:71]: At the moment, its the message returned to the client/sender.   (Note:  not all our returned messages are JSON right now, so I've made this a VARCHAR for now)

Good catch: TASK.task_result is redundant with REPORT_FILE.sending_results_returned.   I need to think about this some...

File: prime-router/docs/proposals/lineage/README.md (hunk 7 above)
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:75]: Not sure I'm a fan of that.   We need a strong way to link data back to what the sender sent, when there are questions or errors.   I think it should be an error if sender fails to send a proper tracking_element_id.

File: prime-router/docs/proposals/lineage/README.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+- might need a report_status in the future.  For example, `wiped` or `archived`, to explain why a blob might have disappeared.
+- The primary key to use for ITEM is tricky.  The sender gives us the `tracking_element`, but we have little control over how robust that is.  Because of this, all queries on `tracking_element` must also filter on report_id.
+- So there's an open question:  should we require global uniqueness of `tracking_elements`?   It would be wonderfully convenient to use `tracking_elements` as primary keys in our database.    However, they would have to be globally unique, because report data can get merged together in arbitrary ways in steps downstream from Receive.
+- There's subtlety with Report parent/child relationships.   Because any given task might take in a bunch of reports, and produce a bunch of reports, there's no guarantee that any of the data in output Report Y came from input Report X.  However we're tracking it in the REPORT_LINEAGE table as if every input report fed every output report.  In practice, its very hard to tell if that's true, in general.
```
[the remark this answers is not in the document]
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:77]: Yes, you could do that.   Its just a question of whether its worth reducing all that glop of per-item data down into some complex many:many mapping at the report level - probably not a high value feature.  As you point out, the right way to do this is for the task could track this as its doing its work.

File: prime-router/docs/proposals/lineage/README.md (hunk 5 above)
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:88]: I plan to write the code to populate the new model without touching the old one.    At some point, that's done, and we can decide whether to deploy to production, or wait until the old style is removed.

Personally, I like the low risk of running in parallel, for a little while.   As long as nothing in production is reading the new tables, they are completely harmless, even if they are erroneous in some way.

File: prime-router/docs/proposals/lineage/README.md (hunk 4 above)
Comment (jimduff-usds) [reply prime-router/docs/proposals/lineage/README.md:46]: At this point, I'm ok either way.    Both require care - its easy to wrap yourself in linguistic spirals.  Like, what do you call the results of a 'send' from the Hub to a PHD?   Its the "received_result", true?    Which sounds like the hub is receiving something....sigh.

**Pull request merged**